keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,573 @@
+
+- Support for [Indexed Jobs](https://git.k8s.io/enhancements/keps/sig-apps/2214-indexed-job)
+- Metrics:
+  - latency
+  - errors
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- E2E test graduates to conformance.
+- Job tracking scales to 10^5 completions.
+
+### Upgrade / Downgrade Strategy
+
+When the feature `JobTrackingWithoutLingeringPods` is enabled for the first
+time, the cluster can have Jobs whose Pods don't have the
+`batch.kubernetes.io/job-completion` finalizer. It would be hard to add the
+finalizer to all Pods while preventing race conditions.
+
+We use `.status.uncountedTerminatedPods != nil` to indicate whether the Job
+was created after the feature was enabled. If this field is nil, the Job
+controller tracks Pods using the legacy tracking.
+
+The kube-apiserver sets `.status.uncountedTerminatedPods` to an empty struct
+when the feature gate `JobTrackingWithoutLingeringPods` is enabled, at Job
+creation. In alpha, apiserver leaves `.status.uncountedTerminatedPods = nil`
+for [Indexed Jobs](https://git.k8s.io/enhancements/keps/sig-apps/2214-indexed-job)
+
+When the feature is disabled after being enabled for some time, the next time
+the Job controller syncs a Job:
+1. It removes finalizers from all Pods owned by the Job.
+2. Sets `.status.uncountedTerminatedPods` to nil.
+
+### Version Skew Strategy
+
+No implications to node runtime.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: JobTrackingWithoutLingeringPods
+    - Components depending on the feature gate:
+      - kube-apiserver
+      - kube-controller-manager
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+
+  Yes.
+  
+  - Removing terminated Pods doesn't affect Job status.
+  - Pods removed by the user or other controllers count towards failures or
+    completions.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  
+  Yes.
+  The job controller removes finalizers in this case.
+  Since some succeeded Pods might have been removed, the job controller will
+  create new Pods to fulfill completions. But this is no different from
+  existing behavior with the legacy tracking.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+
+  Existing Jobs are tracked with legacy Pods.
+
+* **Are there any tests for feature enablement/disablement?**
+
+  Yes, there are integration tests.
```

Comment:
Are there? I'm assuming these are planned, but let's not mention they already exist....

Reply:
reworded